miasm2/ir/analysis.py

```diff
@@ -115,194 +119,246 @@ def graph(self):
-        for ir, _, c_out in zip(irb.irs, irb.c_in, irb.c_out):
+        for k, ir in enumerate(irb.irs):
             j = 0
             while j < len(ir):
-                i_cur = ir[j]
-                if not isinstance(i_cur.dst, ExprId):
-                    pass
-                elif i_cur.dst == self.IRDst:
-                    # never delete irdst
-                    pass
-                elif (isinstance(i_cur.src, ExprOp) and
-                    i_cur.src.op.startswith('call')):
-                    # /!\ never remove ir calls
-                    pass
-                elif i_cur.dst not in c_out:
-                    del(ir[j])
+                cur_instr = ir[j]
+                if (isinstance(cur_instr.dst, ExprId)
+                    and not (irb.label, k, cur_instr) in useful):
+                    del ir[j]
                     modified = True
-                    continue
-                j += 1
+                else:
+                    j += 1
+        return modified
 
-        # print 'state2'
-        # self.dump_bloc_state(irb)
+    def init_useful_instr(self):
+        """Computes a set of triples (bloc, instruction number, instruction)
+        containing initially useful instructions :
+          - Instructions affecting final value of return registers
+          - Instructions affecting IRDst register
+          - Instructions writing in memory
+          - Function call instructions
+        Return set of intial useful instructions
+        """
 
-        return modified
+        useful = set()
+
+        for node in self.g.nodes():
+            if node not in self.blocs:
+                continue
+
+            bloc = self.blocs[node]
+            successors = self.g.successors(node)
+            has_son = bool(successors)
+            for p_son in successors:
+                if p_son not in self.blocs:
+                    # Leaf has lost its son: don't remove anything
+                    # reaching this bloc
+                    for r in self.ira_regs_ids():
+                        useful.update(bloc.cur_reach[-1][r].union(
+                                bloc.defout[-1][r]))
+
+            # Function call, memory write or IRDst affectation
+            for k, ir in enumerate(bloc.irs):
+                for i_cur in ir:
+                    if i_cur.is_function_call():
+                        # /!\ never remove ir calls
+                        useful.add((bloc.label, k, i_cur))
+                    if (isinstance(i_cur.dst, ExprMem)):
+                        useful.add((bloc.label, k, i_cur))
+                    useful.update(bloc.defout[k][self.IRDst])
+
+            # Affecting return registers
+            if not has_son:
+                for r in self.get_out_regs(bloc):
+                    useful.update(bloc.defout[-1][r]
+                                  if bloc.defout[-1][r] else
+                                  bloc.cur_reach[-1][r])
+
+        return useful
+
+    def mark_useful_code(self):
```

Comment:
If this method is only used in the internal dead code algorithm, please mark it as protected (`_mark_useful_code`).